rust: delay: add `coarse_sleep` examples and tests

ojeda opened the pull request:
While I was reviewing #823 I wrote a few `assert!`s to double-check I understood the logic, so why not add them as examples and tests?

We could avoid having to keep these in sync by making the conversion function public and moving the tests as `# Examples`, but I am not sure whether we want to expose another function. So, for the moment, at least add them as tests.

alex commented on rust/kernel/delay.rs:
````diff
@@ -27,6 +21,52 @@ pub fn coarse_sleep(duration: Duration) {
+/// ```
+/// # use core::time::Duration;
+/// # use kernel::delay::coarse_sleep;
+/// coarse_sleep(Duration::new(0, 0));          // Equivalent to `msleep(0)`.
````
These may be easier to read if written:
```suggestion
/// coarse_sleep(Duration::from_secs(0));          // Equivalent to `msleep(0)`.
```

ojeda replied:
I considered that, as well as using `ZERO` etc., but I wanted to keep all the examples the same so that it was easy to tell the "jump" from one ns to the next as well as from one milli to the next (in the `msleep` equivalent), if that makes sense.

Also, given there are several examples plus the equivalence in milliseconds in the comment, it should be fairly clear what the arguments are (i.e. seconds and nanoseconds).

ojeda replied:
On the other hand, people may copy-paste this without knowing there are better constructors, so perhaps it is worth doing it for the examples (but keeping the `#[test]`s with `new` for the reasons above).

ojeda replied:
After thinking about it, I agree it is best to show the friendlier constructors if only because those are what people should likely be using in actual driver code etc. And with a few changes in the examples the "jump"s in milliseconds are still obvious.

Thanks @alex!

alex commented:
:+1: code itself LGTM

ojeda commented:
Thanks @alex for the quick review!

I forgot to Cc @jbaublitz (by the way, GitHub does not appear to show you when I type @, probably because you are not part of the organization -- I just sent you the invitation).

ojeda commented:
Went with @alex suggestion after having improved the examples so that they are more clear when using the friendlier constructors, added a couple examples and separated them to make them easier to read, added saturations tests and added a mirror comment to keep things in sync.

jbaublitz commented:
Sorry I couldn't get around to looking at this sooner but this looks reasonable to me!
